File: 120-triangle/triangle.java
class Solution {
    int fn(int row, int col, List<List<Integer>> triangle, int ROWS, Integer[][] memo) {
            // System.out.println(row + " " + col);
            int i = row;
            int j = col;
            if (memo[i][j] != null) {
                return memo[i][j];
            }
            if (row + 1 == ROWS) {
                memo[i][j] = triangle.get(row).get(col);
                return triangle.get(row).get(col);
            }
            else {
                int a = triangle.get(row).get(col) + fn(row + 1, col, triangle, ROWS, memo);
                int b = triangle.get(row).get(col) + fn(row + 1, col + 1, triangle, ROWS, memo);
                memo[i][j] = Math.min(a, b);
                return Math.min(a, b);
            }
        }
    public int minimumTotal(List<List<Integer>> triangle) {
        int ROWS = triangle.size();
        Integer[][] memo = new Integer[triangle.size()][triangle.size()];
        return fn(0, 0, triangle, ROWS, memo);
    }
}

File: 120-triangle/triangle.py
class Solution:
    def minimumTotal(self, triangle: List[List[int]]) -> int:
        ROWS = len(triangle)
        @cache
        def fn(i ,j):
            if i == ROWS - 1:
                return triangle[i][j]
            
            left = triangle[i][j] + fn(i + 1, j)
            right = triangle[i][j] + fn(i + 1, j + 1)

            return min(left, right)
        return fn(0, 0)